tests/models/shadow/commit.p

```diff
@@ -0,0 +1,115 @@
+//
+// The Commit Protocol takes as input a set of produce requests and commits the
+// associated batches. This involves (1) building and uploading the batch data
+// to cloud storage, (2) committing metadata about the object into each
+// partition targetted by the produce requests and (3) acknowledging each client
+// with the final committed offset.
+//
+
+// An L0d object stores payload data from produced batches without offsets.
+// In the model, the position of an entry in an L0d object is used as its
+// identifer in placeholder batches. The values in the L0d object are unique
+// batch identifiers generated by the client. They represent the "content" of
+// the batch, and must not be used by the model except for validation.
+type L0d_object = seq[int];
+
+machine CommitProtocol {
+  var storage: Storage;
+  var requests: seq[produce_request];
+
+  // the L0d object that will be built and uploaded
+  var object: L0d_object;
+  var object_id: int;
+
+  // mapping from offset in object to source request
+  var object_to_request: seq[int];
+  // mapping from append request to offset of data in object
+  var append_to_object: map[int, int];
+
+  // holds the next request id
+  var request_id: int;
+
+  start state Init {
+    entry (input: (storage: Storage, requests: seq[produce_request])) {
+      storage = input.storage;
+      requests = input.requests;
+      goto Commit;
+    }
+  }
+
+  state Commit {
+    entry {
+      build_L0d();
+      upload_L0d();
+      append_placeholders();
+    }
+
+    // Respond to the produce request with the committed offset.
+    on append_response_event do (response: append_response) {
+      var offset: int;
+      var request: produce_request;
+
+      // map the partition append response back the original produce request
+      offset = object_to_request[append_to_object[response.request_id]];
+      request = requests[offset];
+
+      send request.source, produce_response_event, (
+        request_id = request.request_id,
+        offset = response.offset);
+    }
+  }
+
+  // Build the L0d object from the input produce requests. Batches in the L0d
+  // object may have any ordering as long as the object_to_request mapping
+  // maintains the property that for each offset in the object, the value of
+  // object_to_request[offset] is the offset of the source request.
+  fun build_L0d() {
+    var offset: int;
+    var request: produce_request;
+    while (offset < sizeof(requests)) {
+      object += (sizeof(object), requests[offset].batch_id);
+      object_to_request += (sizeof(object_to_request), offset);
```

Comment:
`sizeof(object_to_request) always == offset` always. What's the idea here?

`while (offset < sizeof(request))`: what about making this `while  (req_ix < sizeof(request))` to avoid confusion with kafka offsets and such

Reply:
> sizeof(object_to_request) always == offset always. What's the idea here?

Yes, this is confusing, because as you point out because the values are the same it is unnecessary. It is bit forward looking to support reshuffling the L0d object contents, but that reshuffle is not present in the current implementation.

I've added a more detailed explanation about how this works here https://github.com/redpanda-data/redpanda/pull/23053/commits/8dce8a10de68b808a57bb7a7708e1db77f61e30b

> while (offset < sizeof(request)): what about making this while  (req_ix < sizeof(request)) to avoid confusion with kafka offsets and such

Good idea that is much easier to read, and I think it makes the shuffle index much more obvious too https://github.com/redpanda-data/redpanda/pull/23053/commits/0b1bd21ce8b0ecc8595b56735a3d3ddbc01bdbcb.